cs/src/comm/epoxy-transport/EpoxyListener.cs

```diff
@@ -83,21 +82,59 @@ public override Task StartAsync()
             return TaskExt.CompletedTask;
         }
 
-        public override Task StopAsync()
+        public override async Task StopAsync()
         {
+            // Request that the accept loop stop
             shutdownTokenSource.Cancel();
+
+            // Stop listening. This causes the accept loop's wait for an
+            // incomming socket to fail, which then causes the accept loop to
+            // look at the value of shutdownTokenSource.Token.
             listener.Stop();
 
-            return acceptTask;
+             // Wait for the accept loop to exit. Once it has exited, no new
+            // connections will be accepted, so we can close the outstanding
+            // ones.
+            await acceptTask;
+
+            // Collect the connections to stop.
+            HashSet<EpoxyConnection> connectionsToClose;
+            lock (connectionsLock)
+            {
+                connectionsToClose = new HashSet<EpoxyConnection>(connections);
+                connections.Clear();
+            }
+
+            // Stop all outstanding connections. EpoxyConnection shutdown is
+            // idempotent, so it's safe to call even if someone else called
+            // StopAsync.
+            var connectionShutdownTasks = new Task[connectionsToClose.Count];
+            int idx = 0;
+            foreach (var connection in connectionsToClose)
+            {
+                logger.Site().Debug("{0} stopping connection {1}", this, connection);
+                connectionShutdownTasks[idx] = connection.StopAsync();
+                ++idx;
+            }
+
+            logger.Site().Debug("Waiting for all connection to stop.");
```

Comment:
s/connection/connections

This message might deserve a higher level. I defer to you.